hasjob/views/campaign.py

```diff
@@ -29,222 +92,233 @@ def chart_interval_for(campaign, default='hour'):
+            )).params(interval=interval, timezone=timezone, campaign_id=campaign.id, dismiss_type=CAMPAIGN_ACTION.DISMISS)
 
-    return outfile.getvalue(), 200, {'Content-Type': 'text/plain'}
+        for hour, count in hourly_views:
+            viewdict[hour]['_combined'] = viewdict[hour].setdefault('_combined', 0) + count
+
+        action_names = []
+
+        for action in campaign.actions:
+            action_names.append(action.name)
+            hourly_views = db.session.query('hour', 'count').from_statement(db.text(
+                '''SELECT date_trunc(:interval, campaign_user_action.created_at AT TIME ZONE 'UTC' AT TIME ZONE :timezone) AS hour, COUNT(*) AS count FROM campaign_user_action WHERE action_id=:action_id GROUP BY hour ORDER BY hour;'''
+            )).params(interval=interval, timezone=timezone, action_id=action.id)
+            for hour, count in hourly_views:
+                viewdict[hour][action.name] = count
+
+            hourly_views = db.session.query('hour', 'count').from_statement(db.text(
+                '''SELECT date_trunc(:interval, campaign_anon_user_action.created_at AT TIME ZONE 'UTC' AT TIME ZONE :timezone) AS hour, COUNT(*) AS count FROM campaign_anon_user_action WHERE action_id=:action_id GROUP BY hour ORDER BY hour;'''
+            )).params(interval=interval, timezone=timezone, action_id=action.id)
+            for hour, count in hourly_views:
+                viewdict[hour][action.name] = viewdict[hour].setdefault(action.name, 0) + count
+
+        if viewdict:
+            # Top-off with site-wide user presence (available since 31 Jan 2015 in user_active_at)
+            minhour = g.user.tz.localize(min(viewdict.keys())).astimezone(UTC).replace(tzinfo=None)
+            maxhour = g.user.tz.localize(max(viewdict.keys()) + timedelta(seconds=3599)).astimezone(UTC).replace(tzinfo=None)
+
+            hourly_views = db.session.query('hour', 'count').from_statement(db.text(
+                '''SELECT date_trunc(:interval, user_active_at.active_at AT TIME ZONE 'UTC' AT TIME ZONE :timezone) AS hour, COUNT(DISTINCT(user_active_at.user_id)) AS count FROM user_active_at WHERE user_active_at.active_at >= :min AND user_active_at.active_at <= :max GROUP BY hour ORDER BY hour;'''
+                )).params(interval=interval, timezone=timezone, min=minhour, max=maxhour)
+
+            for hour, count in hourly_views:
+                viewdict[hour]['_site'] = count
+
+            hourly_views = db.session.query('hour', 'count').from_statement(
+                '''SELECT DATE_TRUNC(:interval, event_session.created_at AT TIME ZONE 'UTC' AT TIME ZONE :timezone) AS hour, COUNT(DISTINCT(anon_user_id)) AS count FROM event_session WHERE event_session.anon_user_id IS NOT NULL AND event_session.created_at >= :min AND event_session.created_at <= :max GROUP BY hour ORDER BY hour'''
```

Comment:
Breaking this into multiple lines could make it easier to read?

Reply:
Old code.